lib/task.ml

```diff
@@ -1,33 +1,102 @@
+open EffectHandlers
+open EffectHandlers.Deep
+
 type 'a task = unit -> 'a
 
-type 'a promise = ('a, exn) result option Atomic.t
+type 'a promise_state =
+  Returned of 'a
+| Raised of exn
+| Pending of (unit, unit) continuation list
 
-exception TasksActive
+type 'a promise = 'a promise_state Atomic.t
 
-type task_msg =
-  Task : 'a task * 'a promise -> task_msg
-| Quit : task_msg
+type message =
+  Work of (unit -> unit)
+| Quit
 
 type pool_data = {
   domains : unit Domain.t array;
-  task_chan : task_msg Multi_channel.t;
+  task_chan : message Multi_channel.t;
   name: string option
 }
 
 type pool = pool_data option Atomic.t
 
-let do_task f p =
+let get_pool_data p =
+  match Atomic.get p with
+  | None -> raise (Invalid_argument "pool already torn down")
+  | Some p -> p
+
+let do_task pd f p =
   try
     let res = f () in
-    Atomic.set p (Some (Ok res))
+    match Atomic.exchange p (Returned res) with
+    | Pending l ->
+        List.iter (fun k -> Multi_channel.send pd.task_chan (Work (continue k))) l
```

Comment:
What if the awaiter is from a different pool? We'd execute it on this pool, is that correct?

Reply:
Good catch. The tasks should run to completion on the pool that they were spawned on, not hop between pools. Let me fix that.